internal/server/boltdbstate/job.go

```diff
@@ -1709,6 +1724,125 @@ func (s *State) taskComplete(jobId string) error {
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if err != nil || len(run.Jobs) < 1 {
+		s.log.Error("failed to retrieve pipeline to complete", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	if job.State == pb.Job_ERROR {
+		run.Status = pb.PipelineRun_FAILED
+	} else if job.State == pb.Job_SUCCESS {
+		// If job Id matches last job queued by pipeline
+		if job.Id == run.Jobs[len(run.Jobs)-1].Id {
+			run.Status = pb.PipelineRun_COMPLETED
+			s.log.Trace("pipeline run is complete", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", run.Sequence)
+		}
+	}
+
+	// PipelineRunPut the new state
+	if err := s.PipelineRunPut(run); err != nil {
+		s.log.Error("failed to complete pipeline run", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	return nil
+}
+
+// pipelineAck checks if the referenced job id has a PipelineStep ref associated with it,
+// and if so, progress the PipelineRun state machine.
+func (s *State) pipelineAck(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab pipeline run that triggered the job based on the PipelineTask Ref
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if err != nil || len(run.Jobs) < 1 {
+		s.log.Error("failed to retrieve pipeline to ack", "job", job.Id, "pipeline id", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+	// Update the new pipeline run state
+	run.Status = pb.PipelineRun_RUNNING
+	s.log.Trace("pipeline is running", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", run.Sequence)
+	if err := s.PipelineRunPut(run); err != nil {
+		s.log.Error("failed to ack pipeline run state", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	return nil
+}
+
+// pipelineCancel will look up the referenced job to see if it has a PipelineStep ref
+// associated with it, and if so, progress the pipelineRun state machine.
+func (s *State) pipelineCancel(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab the pipeline run
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if err != nil || len(run.Jobs) < 1 {
```

Comment:
⏳ Is it possible for there to be 0 jobs & `err` to be nil? If so, and if that is the case when this code run, should L1828 return `err` (which would be nil), or should we return a new error if there are 0 jobs?

Reply:
It is possible for there to be 0 jobs and error is nil, if the pipelinerun was initiated without jobs, since that is a valid structure. Jobs are only added after successful queuing.

But this is a good point, we want to cancel a pipeline in any state so this check isn't necessary here, & I updated the other spots to throw an appropriate error. https://github.com/hashicorp/waypoint/pull/3599/commits/e4b4ec8d02724a5376945a1bae02da25bf0d1ea9